### PR TITLE
GPS acceleration compensation for IMU

### PR DESCRIPTION
**Pull request by digitalentity**

### CODE REVIEW

**Comment by brat002** [comment]
> digitalentity modified the milestones: 1.9, 2.0 3 days ago
That's most disappointing comment of this year. What is the reason to move this PR to 2.0?

**Comment by digitalentity** [comment]
Because it's not ready and not working?

File: src/main/common/filter.c [hunk 1]
```diff
@@ -182,6 +182,13 @@ void firFilterInit(firFilter_t *filter, float *buf, uint8_t bufLength, const flo
     firFilterInit2(filter, buf, bufLength, coeffs, bufLength);
 }
 
+void firFilterReset(firFilter_t *filter, const float value)
+{
+    for (int i = 0; i < filter->bufLength; i++) {
+        filter->buf[0] = value;
```
Comment (junwoo091400) [inline src/main/common/filter.c:188]: filter->buf[i] = value

0 => i (Typo)

File: src/main/flight/imu.c [hunk 2]
```diff
@@ -446,20 +465,30 @@ STATIC_UNIT_TESTED void imuUpdateEulerAngles(void)
...
 }
 
-static bool imuCanUseAccelerometerForCorrection(void)
+static bool imuCanUseAccelerometerForCorrection(t_fp_vector * grav)
 {
-    int32_t axis;
-    int32_t accMagnitudeSq = 0;
+    float accMagnitudeSq = sq(grav->V.X) + sq(grav->V.Y) + sq(grav->V.Z);
+    const float nearness = ABS(100.0f - 100.0f * accMagnitudeSq / sq(GRAVITY_CMSS));
+    return (nearness > MAX_ACC_SQ_NEARNESS) ? true : false;
```
Comment (junwoo091400) [inline src/main/flight/imu.c:472]: Should be 'false : true'

Because nearness should be small value if 'true'(Close to GRAVITY_CMSS)

**Comment by shellixyz** [comment]
@digitalentity I rebased this branch on development here: https://github.com/shellixyz/inav/tree/gps-acceleration-comp-rebased

Is it safe to test ?

**Comment by digitalentity** [comment]
Not sure, probably not. We need to record 'gpsAccelInBodyFrame' and check if it is usable. I never had a chance to do that.

If you have the time - log 'gpsAccelInBodyFrame' but disable the part where it's used for IMU correction.

**Comment by shellixyz** [comment]
OK will do that

**Comment by shellixyz** [comment]
I logged 2 flights today with [this code](https://github.com/shellixyz/inav/tree/gps-acceleration-comp-logging-rdy) but it the values do not look right. Field `imuGravBF`. Using the average predictor is probably not the best but I don't think it is the issue. Do you see something I have done wrong ?

[ArrowV1_LOGS_20180331.zip](https://github.com/iNavFlight/inav/files/1865934/ArrowV1_LOGS_20180331.zip)

**Comment by shellixyz** [comment]
I did 2 more flights today during which I logged `gpsAccelInBodyFrame` (gpsAccBF) instead of `imuGravityInBodyFrame` using this code: https://github.com/shellixyz/inav/tree/gps-acceleration-comp-logging-rdy2

[ArrowV1_LOGS_20180401.zip](https://github.com/iNavFlight/inav/files/1866587/ArrowV1_LOGS_20180401.zip)

**Comment by shellixyz** [comment]
I don't understand what is going on. My logging code is most likely wrong because the output is always at one of the extremes.

**Comment by digitalentity** [comment]
@shellixyz ok, I'm finally back and will have a look at your logs in a few days. Thanks

**Comment by shellixyz** [comment]
I did more flights with a different logging code. Using the debug variable instead of adding fields to the BB. It looks like it worked this time.

You can find the code here: https://github.com/shellixyz/inav/tree/gps-acceleration-comp-logging-rdy3.
Here are the log files: [ArrowV1_LOGS_20180509.zip](https://github.com/iNavFlight/inav/files/1989391/ArrowV1_LOGS_20180509.zip)

The gpsAccelBF values do not seem to be usable to me.

**Comment by digitalentity** [comment]
@shellixyz thanks. Yep, there must be a glitch in calculating the `gpsAccelBF`

**Comment by hamishfagg** [comment]
I've always thought that having a 9DOF sensor (with 3-axis mag) allowed you to always keep orientation. Is that not true here or is this a bug when people are not using mags?

**Comment by shellixyz** [comment]
I also thought a mag would help solving the drift issue but it is possibly not precise enough. I'm not sure. Also it would be nice if the drift was fixed for fixed wings not using a mag.

**Comment by shellixyz** [comment]
@digitalentity Acceleration is dv/dt but to me it looks like it is calculated v/dt. Am I mistaken ?

**Comment by digitalentity** [comment]
@IVData the problem about mag is that we can't make assumptions about it direction relatively to horizon. For gravity we know that it points down (orthogonal to level plane). We can't make that assumption with compass - that's why we can't use it to compensate for anything but heading.

@shellixyz that's how differentiating FIR filter works. Delta is calculated inside FIR code based on FIR coefficients.

dv/dt is a 1-st order FIR with [-1/dt; 1/dt] coefficients.

**Comment by hamishfagg** [comment]
@digitalentity 
> the problem about mag is that we can't make assumptions about it direction relatively to horizon

Why is that? Does earth magnetic field change noticeably in direction within fpv-ing distance? I figured you could take a reading on arm and use that as a reference for the rest of the flight.

Sorry to talk about this here, I tried to get you on slack - hit me up there if you'd like.

**Comment by digitalentity** [comment]
I won't be on slack for a while.
Thing is that for gravity we know - if we don't move ourselves, it points
down.
For magnetic field - every single metal structure disrupts the field and
while for heading an error of 5-10 deg is somewhat acceptable, for tilt
it's not.

**Comment by shellixyz** [comment]
@digitalentity I took a deeper look and understand better how it is working now. I made a program to simulate the GPS acceleration calculation from the GPS data in logs based on your code. It seems to be working fine even though I had to add a lot of filtering to get usable values. I don't calculate the z acceleration because the GPS altitude resolution in the CSV generated by blackbox tools is only 1 meter. Here is the result from one of my logs:

gps_accel_sim

**Pull request closed without merging**